ait/gui/static/js/ait/tlm.js

```diff
@@ -379,14 +304,69 @@ class TelemetryStream
         this._emit('close', this)
     }
 
+    getFullPacketStates () {
+        m.request({ url: '/tlm/latest' }).then( (latest) => {
+            console.log('requesting latest tlm')
+            console.log(latest)
+            this._pkt_states = latest['states']
+            this._counters = latest['counters']
+        })
+    }
 
-    onMessage (event) {
-        if ( !(event.data instanceof ArrayBuffer) ) return
+    checkCounter (packetName, counter) {
+        // returns true if counter is as expected, false if not
+        let lastCounter = this._counters[packetName]
+        if ( counter == lastCounter + 1) {
+            return true
+        } else if ( lastCounter == Math.pow(2, 31) - 1 && counter == 0 ) {
+            return true
+        }
+        console.log('counter mismatch: had ' + lastCounter + ' , got ' + counter)
+        return false
+    }
 
-        let uid  = new DataView(event.data, 1, 4).getUint32(0)
-        let data = new DataView(event.data, 5)
-        let defn = this._dict[uid]
+    onMessage (event) {
+        if ( !(typeof event.data == "string") ) return
+
+        let now = Date.now()
+        let data = JSON.parse(event.data)
+        let packet_name = data['packet']
+        let delta = data['data']
+        let dntoeus = data['dntoeus']
+        let counter = data['counter']
+
+        if ( packet_name in this._pkt_states ) {
+            // check counter is as expected and request full packet states if not
+            let gotNextCounter = this.checkCounter(packet_name, counter)
+            if ( !gotNextCounter ) {
+                this.getFullPacketStates()
+            } else {
+                // add delta to current packet state and update counter
+                if ( Object.keys(delta).length !== 0 ) {
+                    console.log('adding delta to pkt state')
+                    for ( var field in delta ) {
```

Comment:
Do we have to do this iteration and reassignment for copying? Could we just do a copy on the object (if we even need that)?

Reply:
Yes because we are not assuming that deltas for all fields are present in `delta` - we need to keep the information on the states of all fields in `_pkt_states` while updating the deltas we did receive from `delta`. (if I'm understanding your question correctly)